clients/src/main/java/org/apache/kafka/common/internals/SecurityManagerCompatibility.java

```diff
@@ -0,0 +1,355 @@
+            doPrivileged = accessController.getDeclaredMethod("doPrivileged", PrivilegedAction.class);
+            getContext = accessController.getDeclaredMethod("getContext");
+            Class<?> accessControlContext = loader.loadClass("java.security.AccessControlContext");
+            Class<?> subject = loader.loadClass(Subject.class.getName());
+            getSubject = subject.getDeclaredMethod("getSubject", accessControlContext);
+            // Note that we use the real Subject.class rather than the `subject` variable
+            // This allows for mocking out the method without changing the argument types.
+            doAs = subject.getDeclaredMethod("doAs", Subject.class, PrivilegedExceptionAction.class);
+        }
+
+        @Override
+        public <T> T doPrivileged(PrivilegedAction<T> action) {
+            try {
+                return (T) doPrivileged.invoke(null, action);
+            } catch (IllegalAccessException e) {
+                throw new UnsupportedOperationException(e);
+            } catch (InvocationTargetException e) {
+                Throwable cause = e.getCause();
+                if (cause instanceof RuntimeException) {
+                    throw (RuntimeException) cause;
+                } else {
+                    throw new RuntimeException(cause);
+                }
+            }
+        }
+
+        /**
+         * @return the result of AccessController.getContext(), of type AccessControlContext
+         */
+        private Object getContext() {
+            try {
+                return getContext.invoke(null);
+            } catch (IllegalAccessException e) {
+                throw new UnsupportedOperationException(e);
+            } catch (InvocationTargetException e) {
+                Throwable cause = e.getCause();
+                if (cause instanceof RuntimeException) {
+                    throw (RuntimeException) cause;
+                } else {
+                    throw new RuntimeException(cause);
+                }
+            }
+        }
+
+        /**
+         * @param context The current AccessControlContext
+         * @return The result of Subject.getSubject(AccessControlContext)
+         */
+        private Subject getSubject(Object context) {
+            try {
+                return (Subject) getSubject.invoke(null, context);
+            } catch (IllegalAccessException e) {
+                throw new UnsupportedOperationException(e);
+            } catch (InvocationTargetException e) {
+                Throwable cause = e.getCause();
+                if (cause instanceof RuntimeException) {
+                    throw (RuntimeException) cause;
+                } else {
+                    throw new RuntimeException(cause);
+                }
+            }
+        }
+
+        @Override
+        public Subject current() {
+            return getSubject(getContext());
+        }
```

Comment:
It'd be a little less verbose if we inlined all of this into a single method; any reason not to?
```suggestion
        @Override
        public Subject current() {
            try {
                // AccessController.getContext()
                Object context = getContext.invoke(null);
                // Subject.getSubject(AccessControlContext)
                return (Subject) getSubject.invoke(null, context);
            } catch (IllegalAccessException e) {
                throw new UnsupportedOperationException(e);
            } catch (InvocationTargetException e) {
                Throwable cause = e.getCause();
                if (cause instanceof RuntimeException) {
                    throw (RuntimeException) cause;
                } else {
                    throw new RuntimeException(cause);
                }
            }
        }
```

Reply:
This was in order to separate the methods doing the reflection, and the methods mapping between the new and old APIs.

Deduping the try-catch has already made this a lot more concise, so LMK if you think we should still remove the getSubject/getContext/doAs methods.